lib/parsers/package-lock-parser.ts

```diff
@@ -49,66 +64,240 @@ export class PackageLockParser implements LockfileParser {
+      for (const edge of outEdges) {
+        depGraph.setEdge(edge.v, newNode);
+      }
+    }
+
+    // Once completed for all nodes in a cycle, original cyclic nodes can
+    // be removed.
+    for (const start of cycle) {
+      depGraph.removeNode(start);
+    }
+
+    return cycleStarts;
+  }
+
+  private createGraphOfDependencies(depMap: DepMap): graphlib.Graph {
+    const depGraph = new graphlib.Graph();
+    for (const depName of Object.keys(depMap)) {
+      depGraph.setNode(depName, depName);
+    }
+    for (const [depPath, dep] of _.entries(depMap)) {
+      for (const depName of dep.requires) {
+        const subDepPath = this.findDepsPath(depPath, depName, depMap);
+        // direction is from the dependency to the package requiring it
+        depGraph.setEdge(subDepPath, depPath);
+      }
+    }
+
+    return depGraph;
+  }
 
-    // try to get list of deps on the path
-    const deps: PackageLockDeps = _.get(lockFile, lockfilePath);
-    const dep: PackageLockDep = _.get(deps, depName);
-    // If exists and looked-up dep is there
-    if (dep) {
-      // update the tree
-      depSubTree.version = dep.version;
-      depSubTree.depType = (isDev || dep.dev) ? DepType.dev : DepType.prod;
-      // check if we already have a package at particular version in the traversed path
-      const depKey = `${depName}@${dep.version}`;
-      if (depPath.indexOf(depKey) >= 0) {
-        depSubTree.cyclic = true;
-      } else {
-        // if not, add it
-        depPath.push(depKey);
-        // repeat the process for dependencies of looked-up dep
-        const newDeps = dep.requires ? Object.keys(dep.requires) : [];
-
-        await Promise.all(newDeps.map(async (subDep) => {
-          depSubTree.dependencies[subDep] = await this.buildSubTreeRecursiveFromPackageLock(
-            subDep, [...lockfilePath, depName, 'dependencies'], lockFile, depPath.slice(), isDev);
-        }));
+  // dependency in package-lock.json v1 can be defined either inside `dependencies`
+  // of other dependency or anywhere upward towards root
+  private findDepsPath(startPath: string, depName: string, depMap: DepMap): string {
+    const depPath = startPath.split(this.pathDelimiter);
+    while (depPath.length) {
+      const currentPath = depPath.concat(depName).join(this.pathDelimiter);
+      if (depMap[currentPath]) {
+        return currentPath;
       }
-      return depSubTree;
-    } else {
-      // tree was walked to the root and dependency was not found
-      if (!lockfilePath.length) {
-        throw new Error(`Dependency ${depName} was not found in package-lock.json.
-          Your package.json and package-lock.json are probably out of sync.
-          Please run "npm install" and try again.`);
+      depPath.pop();
+    }
+
+    if (!depMap[depName]) {
+      throw new Error(`Dependency ${depName} was not found in package-lock.json.
+        Your package.json and package-lock.json are probably out of sync.
```

Comment:
when playing with some packages, this error happened to me not because the files were out-of-sync, but because one transitive dep failed to install. So this can happen sometimes not related to top-level deps. FYI